src/s3/S3Controller.ts

```diff
@@ -0,0 +1,155 @@
+import { fromEnv } from '@aws-sdk/credential-providers';
+import {
+  CopyObjectCommand,
+  DeleteObjectCommand,
+  HeadObjectCommand,
+  ListObjectsV2Command,
+  PutObjectCommand,
+  S3Client,
+} from '@aws-sdk/client-s3';
+
+export default class S3Controller {
+  private client: S3Client;
+  private publicFileLinkPrefix: string;
+  private bucket: string;
+
+  constructor() {
+    this.publicFileLinkPrefix = process.env.S3_PUBLIC_FILE_LINK_PREFIX;
+    this.bucket = process.env.S3_BUCKET_NAME;
+
+    this.client = new S3Client({
+      region: process.env.S3_REGION,
+      endpoint: process.env.S3_ENDPOINT,
+      credentials: fromEnv(),
+    });
+  }
+
+  private async objectExists(key: string) {
+    try {
+      const headCommand = new HeadObjectCommand({
+        Bucket: this.bucket,
+        Key: key,
+      });
+
+      await this.client.send(headCommand);
+      return true;
+    } catch (e) {
+      if (e.$metadata.httpStatusCode === 404) {
+        return false;
+      } else {
+        throw e;
+      }
+    }
+  }
+
+  public async listObjectKeys(prefixes: string[]) {
+    // Note ListObjects command only returns up to 1000 objects
+    // Need to update code if this exceeds
+    let keys: string[] = [];
+
+    for (const prefix of prefixes) {
+      const command = new ListObjectsV2Command({
+        Bucket: this.bucket,
+        Prefix: prefix,
+      });
+
+      const s3Objects = await this.client.send(command);
+
+      if (s3Objects.KeyCount !== 0) {
+        s3Objects.Contents.forEach((object) => {
+          keys.push(object.Key);
+        });
+      }
+    }
+
+    return keys;
+  }
+
+  public async uploadFile(key: string, fileContent: Buffer, contentType: string) {
+    if (await this.objectExists(key)) {
+      throw new Error(`Failed to upload to s3. File already exists: '${key}'.`);
+    }
+
+    const command = new PutObjectCommand({
+      Bucket: this.bucket,
+      Key: key,
+      Body: fileContent,
+      ContentType: contentType,
+    });
+
+    await this.client.send(command);
+
+    console.log(`Successfully uploaded file to s3: ${key}`);
+
+    return `${this.publicFileLinkPrefix}${key}`;
+  }
+
+  public async deleteFile(link: string) {
+    const splitLink = this.splitLink(link);
```

Comment:
Up to you if you wanna change this but maybe
```suggestion
    const linkDetails = this.parseLink(link);
```